# Task 1 - Fitbit Dataset

1. Read this dataset in pandas , mysql and mongodb 
2. while creting a table in mysql dont use manual approach to create it  ,always use a automation to create a table in mysql

 #### hint - use csvkit library to automate this task and to load a data in bulk in you mysql 
3. convert all the dates avaible in dataset to timestamp format in pandas and in sql you to convert it in date format

4. Find out in this data that how many unique id's we have 

5. which id is one of the active id that you have in whole dataset 

6. how many of them have not logged there activity find out in terms of number of ids 

7. Find out who is the laziest person id that we have in dataset 

8. Explore over an internet that how much calories burn is required for a healthy person and find out how many healthy person we have in our dataset 

9. how many person are not a regular person with respect to activity try to find out those 

10. who is the thired most active person in this dataset find out those in pandas and in sql both . 

11. who is the 5th most laziest person avilable in dataset find it out 

12. what is a totla acumulative calories burn for a person find out

In [1]:
import pandas as pd
import pymongo
import mysql.connector as conn
from sqlalchemy import engine
import json

# 1. Read this dataset in pandas , mysql and mongodb

In [76]:
fitbit_df = pd.read_csv('FitBit data.csv')

In [20]:
fitbit_df

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0    1503960366    3/25/2016       11004       7.110000         7.110000   
1    1503960366    3/26/2016       17609      11.550000        11.550000   
2    1503960366    3/27/2016       12736       8.530000         8.530000   
3    1503960366    3/28/2016       13231       8.930000         8.930000   
4    1503960366    3/29/2016       12041       7.850000         7.850000   
..          ...          ...         ...            ...              ...   
452  8877689391     4/8/2016       23014      20.389999        20.389999   
453  8877689391     4/9/2016       16470       8.070000         8.070000   
454  8877689391    4/10/2016       28497      27.530001        27.530001   
455  8877689391    4/11/2016       10622       8.060000         8.060000   
456  8877689391    4/12/2016        2350       1.780000         1.780000   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                         0.0                2.57                      0.46   
1                         0.0                6.92                      0.73   
2                         0.0                4.66                      0.16   
3                         0.0                3.19                      0.79   
4                         0.0                2.16                      1.09   
..                        ...                 ...                       ...   
452                       0.0               11.10                      0.63   
453                       0.0                0.00                      0.02   
454                       0.0               21.92                      1.12   
455                       0.0                1.47                      0.15   
456                       0.0                0.00                      0.00   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                   4.07                     0.00                 33   
1                   3.91                     0.00                 89   
2                   3.71                     0.00                 56   
3                   4.95                     0.00                 39   
4                   4.61                     0.00                 28   
..                   ...                      ...                ...   
452                 8.62                     0.00                 70   
453                 8.02                     0.00                 90   
454                 4.46                     0.00                128   
455                 6.37                     0.01                 18   
456                 1.78                     0.00                  0   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                     12                   205               804      1819  
1                     17                   274               588      2154  
2                      5                   268               605      1944  
3                     20                   224              1080      1932  
4                     28                   243               763      1886  
..                   ...                   ...               ...       ...  
452                   29                   359               982      4196  
453                    9                   289              1052      3841  
454                   46                   211              1055      4526  
455                    7                   225              1190      2820  
456                    0                    58               531       938  

[457 rows x 15 columns]

In [21]:
fitbit_df.dtypes

Id                            int64
ActivityDate                 object
TotalSteps                    int64
TotalDistance               float64
TrackerDistance             float64
LoggedActivitiesDistance    float64
VeryActiveDistance          float64
ModeratelyActiveDistance    float64
LightActiveDistance         float64
SedentaryActiveDistance     float64
VeryActiveMinutes             int64
FairlyActiveMinutes           int64
LightlyActiveMinutes          int64
SedentaryMinutes              int64
Calories                      int64
dtype: object

In [22]:
fitbit_df.to_json('fitbit.json',orient='records')

In [6]:
# Inserting Data into Mongo DB
client = pymongo.MongoClient("mongodb+srv://bhoodev:th1nk11t@bhoodev.angnq.mongodb.net/?retryWrites=true&w=majority")
mongo_database = client["live_class_task_31_july"]
collection = mongo_database['fitbit']
with open('fitbit.json') as file:
    fitbit_json = json.load(file)
collection.insert_many(fitbit_json)

# 2. While creating a table in mysql don't use manual approach to create it  ,always use a automation to create a table in mysql
### hint - use csvkit library to automate this task and to load a data in bulk in you mysql 

In [23]:
mysql_settings = {
"host":"localhost",
"user":"root",
"passwd":"th1nk11t",
"allow_local_infile":True
}

In [24]:
mysql = conn.Connect(**mysql_settings)
cursor = mysql.cursor()

In [25]:
cursor.execute('CREATE DATABASE IF NOT EXISTS liveclasstask31july')

In [26]:
mysql_engin = engine.create_engine("mysql+pymysql://root:th1nk11t@localhost/liveclasstask31july")

In [11]:
fitbit_df.to_sql('fitbit',con=mysql_engin)

ValueError: Table 'fitbit' already exists.

# 3. convert all the dates avaible in dataset to timestamp format in pandas and in sql you to convert it in date format

In [27]:
fitbit_df

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0    1503960366    3/25/2016       11004       7.110000         7.110000   
1    1503960366    3/26/2016       17609      11.550000        11.550000   
2    1503960366    3/27/2016       12736       8.530000         8.530000   
3    1503960366    3/28/2016       13231       8.930000         8.930000   
4    1503960366    3/29/2016       12041       7.850000         7.850000   
..          ...          ...         ...            ...              ...   
452  8877689391     4/8/2016       23014      20.389999        20.389999   
453  8877689391     4/9/2016       16470       8.070000         8.070000   
454  8877689391    4/10/2016       28497      27.530001        27.530001   
455  8877689391    4/11/2016       10622       8.060000         8.060000   
456  8877689391    4/12/2016        2350       1.780000         1.780000   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                         0.0                2.57                      0.46   
1                         0.0                6.92                      0.73   
2                         0.0                4.66                      0.16   
3                         0.0                3.19                      0.79   
4                         0.0                2.16                      1.09   
..                        ...                 ...                       ...   
452                       0.0               11.10                      0.63   
453                       0.0                0.00                      0.02   
454                       0.0               21.92                      1.12   
455                       0.0                1.47                      0.15   
456                       0.0                0.00                      0.00   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                   4.07                     0.00                 33   
1                   3.91                     0.00                 89   
2                   3.71                     0.00                 56   
3                   4.95                     0.00                 39   
4                   4.61                     0.00                 28   
..                   ...                      ...                ...   
452                 8.62                     0.00                 70   
453                 8.02                     0.00                 90   
454                 4.46                     0.00                128   
455                 6.37                     0.01                 18   
456                 1.78                     0.00                  0   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                     12                   205               804      1819  
1                     17                   274               588      2154  
2                      5                   268               605      1944  
3                     20                   224              1080      1932  
4                     28                   243               763      1886  
..                   ...                   ...               ...       ...  
452                   29                   359               982      4196  
453                    9                   289              1052      3841  
454                   46                   211              1055      4526  
455                    7                   225              1190      2820  
456                    0                    58               531       938  

[457 rows x 15 columns]

In [88]:
fitbit_df.dtypes

Id                                   int64
ActivityDate                datetime64[ns]
TotalSteps                           int64
TotalDistance                      float64
TrackerDistance                    float64
LoggedActivitiesDistance           float64
VeryActiveDistance                 float64
ModeratelyActiveDistance           float64
LightActiveDistance                float64
SedentaryActiveDistance            float64
VeryActiveMinutes                    int64
FairlyActiveMinutes                  int64
LightlyActiveMinutes                 int64
SedentaryMinutes                     int64
Calories                             int64
dtype: object

In [90]:
fitbit_df['ActivityDate'] = pd.to_datetime(fitbit_df['ActivityDate'])

#### droping and creating again fitbit again

In [59]:
cursor.execute('USE liveclasstask31july')
cursor.execute('DROP TABLE IF EXISTS fitbit')
fitbit_df.to_sql('fitbit',con=mysql_engin,index=False)

457

In [60]:
columns_to_timestamp = ['VeryActiveMinutes','FairlyActiveMinutes','LightlyActiveMinutes','SedentaryMinutes']

In [61]:
for column in columns_to_timestamp:
    query = f"update fitbit set {column} = timestamp(ActivityDate,convert(CONCAT(cast(floor({column}/60) AS CHAR),':',cast({column}% 60 AS CHAR),':0'),time))"
    cursor.execute(query)
    query2 = f"alter table fitbit modify {column} TIMESTAMP"
    cursor.execute(query2)

In [62]:
mysql.commit()

In [63]:
cursor.execute("DESC fitbit")
cursor.fetchall()

[('Id', b'bigint', 'YES', '', None, ''),
 ('ActivityDate', b'datetime', 'YES', '', None, ''),
 ('TotalSteps', b'bigint', 'YES', '', None, ''),
 ('TotalDistance', b'double', 'YES', '', None, ''),
 ('TrackerDistance', b'double', 'YES', '', None, ''),
 ('LoggedActivitiesDistance', b'double', 'YES', '', None, ''),
 ('VeryActiveDistance', b'double', 'YES', '', None, ''),
 ('ModeratelyActiveDistance', b'double', 'YES', '', None, ''),
 ('LightActiveDistance', b'double', 'YES', '', None, ''),
 ('SedentaryActiveDistance', b'double', 'YES', '', None, ''),
 ('VeryActiveMinutes', b'timestamp', 'YES', '', None, ''),
 ('FairlyActiveMinutes', b'timestamp', 'YES', '', None, ''),
 ('LightlyActiveMinutes', b'timestamp', 'YES', '', None, ''),
 ('SedentaryMinutes', b'timestamp', 'YES', '', None, ''),
 ('Calories', b'bigint', 'YES', '', None, '')]

#### Converting minutes into time delta to perform operations on them

In [14]:
fitbit_df['VeryActiveMinutes'] = pd.to_timedelta(fitbit_df['VeryActiveMinutes'],unit='m')
fitbit_df['FairlyActiveMinutes'] = pd.to_timedelta(fitbit_df['FairlyActiveMinutes'],unit='m')
fitbit_df['LightlyActiveMinutes'] = pd.to_timedelta(fitbit_df['LightlyActiveMinutes'],unit='m')
fitbit_df.dtypes

Id                                    int64
ActivityDate                 datetime64[ns]
TotalSteps                            int64
TotalDistance                       float64
TrackerDistance                     float64
LoggedActivitiesDistance            float64
VeryActiveDistance                  float64
ModeratelyActiveDistance            float64
LightActiveDistance                 float64
SedentaryActiveDistance             float64
VeryActiveMinutes           timedelta64[ns]
FairlyActiveMinutes         timedelta64[ns]
LightlyActiveMinutes        timedelta64[ns]
SedentaryMinutes                      int64
Calories                              int64
dtype: object

# 4. Find out in this data that how many unique id's we have 

In [16]:
len(fitbit_df['Id'].unique())

35

# 5. which id is one of the active id that you have in whole dataset 

In [17]:
fitbit_df.groupby('Id')['Calories'].mean().sort_values(ascending=False)

Id
8877689391    3451.166667
8378563200    3356.166667
5577150313    3300.272727
4020332650    3075.375000
1644430081    2916.400000
8053475328    2893.181818
4702921684    2821.466667
6775888955    2724.777778
7007744171    2627.083333
2022484408    2475.333333
8583815059    2391.250000
2891001357    2273.375000
1927972279    2254.000000
7086361926    2177.000000
6290855005    2165.600000
4445114986    2107.800000
6117666160    2099.300000
6962181067    2089.214286
8792009665    2073.666667
2347167796    2021.333333
4319703577    1994.250000
4057192912    1903.531250
3372868164    1860.300000
4558609924    1830.166667
4388161847    1805.250000
5553957443    1802.500000
1503960366    1796.210526
6391747486    1763.111111
2873212765    1695.750000
1844505072    1615.916667
2320127002    1532.083333
8253242879    1463.000000
3977333714    1398.083333
2026352035    1355.500000
1624580081    1352.894737
Name: Calories, dtype: float64

# 6. how many of them have not logged there activity find out in terms of number of ids 

In [18]:
# person who use to covert the distance but they donot log it
len(fitbit_df['Id'].unique()) - len(fitbit_df[fitbit_df['LoggedActivitiesDistance'] != 0]['Id'].drop_duplicates())

29

### There are total 29 person who NEVER log there activity

# 7. Find out who is the laziest person id that we have in dataset 

In [19]:
 fitbit_df.groupby('Id')['Calories'].mean().sort_values().head(1)

Id
1624580081    1352.894737
Name: Calories, dtype: float64

### Person with Id = 1624580081 is the laziest person in the dataset

# 8. Explore over an internet that how much calories burn is required for a healthy person and find out how many healthy person we have in our dataset 

In [20]:
# Reference from https://www.goodto.com/wellbeing/diets-exercise/what-is-calorie-how-many-lose-weigt-425557
# The average person burns around 1800 calories a day doing absolutely nothing.
fitbit_df[fitbit_df['Calories'] >= 1800]["Id"].drop_duplicates()

0      1503960366
38     1644430081
48     1844505072
60     1927972279
72     2022484408
101    2320127002
108    2347167796
125    2873212765
135    2891001357
143    3372868164
165    4020332650
199    4057192912
229    4319703577
241    4388161847
249    4445114986
264    4558609924
276    4702921684
291    5553957443
303    5577150313
314    6117666160
324    6290855005
334    6391747486
343    6775888955
352    6962181067
366    7007744171
378    7086361926
390    8053475328
408    8253242879
413    8378563200
425    8583815059
433    8792009665
445    8877689391
Name: Id, dtype: int64

In [21]:
len(fitbit_df[fitbit_df['Calories'] >= 1800]["Id"].drop_duplicates())

32

### Total 32 persons are fall into healty category because they on an average burn 1800 Calories

# 9. how many person are not a regular person with respect to activity try to find out those 

In [22]:
fitbit_df.groupby('Id').describe()['Calories']['std'].sort_values(ascending=False).head(1)

Id
6775888955    1153.500951
Name: std, dtype: float64

#### Id = 6775888955 is most irregular person as we can he his standard deviation maximum as comare to other person

# 10. who is the thired most active person in this dataset find out those in pandas and in sql both . 

In [23]:
fitbit_df.groupby('Id')['Calories'].mean().sort_values(ascending=True).tail(3).head(1)

Id
5577150313    3300.272727
Name: Calories, dtype: float64

#### Id = 5577150313 is the Third most active person

# 11. who is the 5th most laziest person avilable in dataset find it out 

In [24]:
fitbit_df.groupby('Id')['Calories'].mean().sort_values().head(5).tail(1)

Id
2320127002    1532.083333
Name: Calories, dtype: float64

### 5th most laziest persons is with Id = 2320127002

# 12. what is a total acumulative calories burn for a person find out

In [25]:
fitbit_df.groupby('Id')['Calories'].sum()

Id
1503960366    34128
1624580081    25705
1644430081    29164
1844505072    19391
1927972279    27048
2022484408    29704
2026352035    16266
2320127002    18385
2347167796    30320
2873212765    20349
2891001357    18187
3372868164    18603
3977333714    16777
4020332650    98412
4057192912    60913
4319703577    23931
4388161847    14442
4445114986    31617
4558609924    21962
4702921684    42322
5553957443    21630
5577150313    36303
6117666160    20993
6290855005    21656
6391747486    15868
6775888955    24523
6962181067    29249
7007744171    31525
7086361926    26124
8053475328    31825
8253242879    17556
8378563200    40274
8583815059    19130
8792009665    24884
8877689391    41414
Name: Calories, dtype: int64